src/osdc/Objecter.cc

```diff
@@ -1328,15 +1328,6 @@ void Objecter::handle_osd_map(MOSDMap *m)
   for (list<LingerOp*>::iterator p = need_resend_linger.begin();
        p != need_resend_linger.end(); ++p) {
     LingerOp *op = *p;
-    if (!op->session) {
```

Comment:
So ```_recalc_linger_op_target()``` invokes ```_session_linger_op_assign()```, but only if ```_get_session()``` had a different answer than previously — and one of the allowed answers is "homeless". It looks to me like a newly-homeless op can still be put on the ```need_resend_linger``` list?
Although since this code doesn't handle that either, maybe it's not the case. Can we at least add an ```assert(session);```?

Reply:
`_recalc_linger_op_target` only gets called by `Objecter::_scan_requests` which assures the checked linger op always has a `OSDSession` associated with it.

> It looks to me like a newly-homeless op can still be put on the `need_resend_linger` list?

yes, it's perfectly valid, so in `Objecter::handle_osd_map` before resending the linger op we have a homeless check.

> Can we at least add an `assert(session);?`

sure :)